docs/design/coreclr/botr/data_contracts_for_coreclr.md

```diff
@@ -0,0 +1,430 @@
+            public virtual int BaseSize(TargetPointer methodTablePointer) { return _target.ReadInt32(_target.GetTargetPointerForField(methodTablePointer, _baseSizeField)); }
+        }
+
+        // This is used for version 2 of the contract, where the dynamic type id is no longer present
+        class MethodTableContract_2 : IAlgorithmContract
+        {
+            Target _target;
+            FieldLayout _baseSizeField;
+
+            public MethodTableContract_2(Target target) { _target = target; }
+
+            void IAlgorithmContract.Init()
+            {
+                _baseSizeField = _target.Contract.GetFieldLayout("MethodTable", "baseSize");
+                if (_baseSizeField.Type != FieldType.Int32Type)
+                    throw new InvalidContractSpec();
+            }
+
+            public virtual int DynamicTypeID(TargetPointer methodTablePointer) { throw new NotImplementedException(); }
+
+            public virtual int BaseSize(TargetPointer methodTablePointer) { return _target.ReadInt32(_target.GetTargetPointerForField(methodTablePointer, _baseSizeField)); }
+        }
+    }
+}
+```
+
+Within the contract algorithm specification, the intention is that all interesting api work is done by using an instance of the `Target` class.
+
+## Arrangement of contract specifications in the repo
+
+Specs shall be stored in the repo in a set of directories. `docs/design/datacontracts`
+
+### Global Value Contracts
+For global value contracts they shall be stored in `docs/design/datacontracts/globals/<contract_name>/<version>`
+
+The name of each contract shall be `docs/design/datacontracts/globals/<contract_name>/<contract_name>_<version>.md`
+
+The format of each contract spec shall be
+
+
+```
+# Contract '<contract_name>' Version '<version_number>'
+
+Insert description of contract, and what its for here.
+
+## Values
+| Global Name | Type | Value |
+| --- | --- | --- |
+| SomeGlobal | Int32 | 1 |
+| SomeOtherGlobal | Int8 | 0 |
+```
+
+Which should format like:
+# Contract '<contract_name>' Version '<version_number>'
+
+Insert description of contract, and what its for here.
+
+## Values
+| Global Name | Type | Value |
+| --- | --- | --- |
+| SomeGlobal | Int32 | 1 |
+| SomeOtherGlobal | Int8 | 0 |
+
+
+### Data Structure Contracts
+For global value contracts they shall be stored in `docs/design/datacontracts/datalayout/<contract_name>/<version>`
+
+The name of each contract shall be `docs/design/datacontracts/datalayout/<contract_name>/<contract_name>_<version>.md`
+
+The format of each contract spec shall be
+
+
+```
+# Contract '<type_name>_layout' Version '<version_number>'
+
+Insert description of type, and what its for here.
+
+## Values
+
+Type name '<type_name>'
+
+| Field Name | Type | Offset |
+| --- | --- | --- |
+| FirstField | Int32 | 0 |
+| SecondField | Int8 | 4 |
```

Comment:
I don't understand what the "offset" here refers to.  The offsets are something we will discover when a reader begins to inspect a particular target.  What possible offset value can we write down if the data structure has native pointers or native pointer-sized values?

Reply:
The thought was that different bitnesses would have different versions. My hope was that these data struct layout contracts would not be directly referenced in source code of the runtime ever, but chosen via some sort of macro/constexpr/something horror that was aware of the possible options. Perhaps I should prototype that, and see how bad of an idea it is.

Reply:
> The thought was that different bitnesses would have different versions. My hope was that these data struct layout contracts would not be directly referenced in source code of the runtime ever, but chosen via some sort of macro/constexpr/something horror that was aware of the possible options. Perhaps I should prototype that, and see how bad of an idea it is.

I strongly feel that having multiple sources of truth is a bad idea.  The structure offsets should be queried from the target runtime.  They should not be stored as artifacts in the repo.  


Moreover, I think this creates challenges for being able to support custom builds / community builds with a single reader - if they need to open a PR to `main` to check in a new version of a data structure layout contract, that means the community port isn't really usable with an extant reader.  (Or possibly I don't understand how contract versioning will work in this case)

Reply:
I think defining default offsets in documentation and providing a standard mechanism for contracts to encode an alternate value explicitly would be a nice tradeoff. Encoding every field offset in memory could be quite sizable and most of the offsets never change, but not having any way to adjust them without a version bump would be pretty restrictive.

Reply:
I agree.

This also means that we need to have an automated way to compare the actual offsets with the baseline and only include the ones that need overriding. Doing this manually would be very error prone and labor intensive.

Reply:
> Encoding every field offset in memory could be quite sizable and most of the offsets never change, but not having any way to adjust them without a version bump would be pretty restrictive.

I disagree with this concept a great deal. I appreciate the size concerns here and want to reach a compromise because I see how important it is. However, I think we should err on the side of understand the breadth of what we need and size costs prior to creating an implied system that will be very hard to reason about without tooling alluded to by @jkotas. This, to me, is premature in the sense that we can easily remove data from a system if we decide we don't need it, but it is very painful to retroactively go back and add in all the constants we litter across the system.

Reply:
I've seen a few approaches for this in the past:
1. Every time we add new fields with expected offsets we added static asserts to the code verifying the offset doesn't change.
2. For a while we had a tool that extracted type offsets from PDBs so we compare DAC type definitions to runtime type definitions. It was designed to assume that any same named field was shared by default which meant there was no explicit opt-in for new fields, but also plenty of false positives when the tool incorrectly assumed new fields were shared. Ultimately the tool fell into disrepair as the team did not believe it found enough issues to justify the ongoing maintainance.
3. Testing, often at feature/integration level, would start failing if the DAC was reading data from incorrect offsets.

The most appealing option to me initially would be to explore (1) and use (3) as an added layer of defense. If the static assert ever fires we could decide either to replace the assert with an explicit encoding of the field offset, roll back the offset change if it was unintended, or do a breaking change to the contract and update the doc+assert with a new expected offset.

I'm certainly open to other alternatives if folks want to suggest them.

> However, I think we should err on the side of understand the breadth of what we need and size costs

If you want a rough estimate I'm guessing 50+ types and 300+ unique fields. You might be able to make a better estimate looking at all the types and fields in the current DAC PDB, but its probably pretty laborious to separate the fields which happen to be defined on those types vs. the ones that are actually read from.

Reply:
> I'm certainly open to other alternatives if folks want to suggest them.

An alternative that @AaronRobinsonMSFT, @davidwrighton and I came up earlier today:

- Create a static const array with the offsets relevant to data contract in the runtime binary: `static const int[] = { <unique sequence that identifies this array>, offsetof(TypeA, FieldX), offsetof(TypeA, FieldY), offsetof(TypeB, FieldZ), .... }` 

- Run a post-processing tool on the runtime binary that:
   - Finds the array of offsets in the runtime binary using the unique sequence
   - Compares the offsets with the baseline contract
   - Produces compact delta representation of the contract and blasts it into the runtime binary (it can even blast it over the array of offsets used as an input to save space). 

This approach avoids the need to build a symbol parser that is complicated and fragile, and it gives us full flexibility in creating a compact contract representation from multiple data sources without incurring runtime costs. One of the data sources can be the managed part of the runtime that addresses @MichalStrehovsky 's [feedback](https://github.com/dotnet/runtime/pull/99936#discussion_r1529817476).

Reply:
I'm not sure I've followed properly, can you help me correct my understanding? 
The static const array is being defined as part of the runtime binary so we'd expect binary size to increase by the size of this array. This array would include a position for every field ever referenced in the data contract so perhaps ~500 fields, 4 byte offsets = ~2KB. Lets say new compressed version is ~200 bytes. Is something going to reclaim the other 1800 bytes and shrink the binary again? If not I'm not sure what benefit we gained from the compression step?

For the multiple data source parts was that meaning that some offsets will just be zero filled initially to reserve space because we can't use a C++ offsetof() expression to compute them?

Reply:
> Is something going to reclaim the other 1800 bytes and shrink the binary again?

The original idea was to just leave these 1800 unused bytes in the binary (it is not that big of a deal to have never used 1800 bytes in the binary). As @lambdageek suggested, we can do better by running the tool on the object file instead.

> the multiple data source part

From the discussion so far, I think we want to combine these 3 data sources to produce the final contract descriptor embedded in the runtime binary:
- The actual offsets of unmanaged data structures laid out by C++ compiler 
- The actual offsets of managed data structures laid out by AOT compiler
- The baseline contract descriptor that contains default implied offset values (we want to exclude the offsets that match the baseline from the embedded contract descriptor to reduce its size)

> we run the tool on an object file (or multiple files) that just has the magic const

I like the idea. It would address the @noahfalk concern about unused space in the binary. It would also work better for the creating the contract 

> I'm not sure if native linkers will keep an unreferenced static const array

Yes, it is not unusual to run into this issue for data structures that are only used out-of-proc - for example, https://github.com/dotnet/runtime/blob/e12e2fa6cbdd1f4b0c8ad1b1e2d960a480c21703/src/coreclr/vm/util.cpp#L1507-L1516 . We typically solve it by adding volatile read/write or a similar measure that the compiler is not able to optimize out.

Reply:
I started a little prototype mostly to understand how the build integration will look.  It's not too bad.
 https://github.com/lambdageek/magic-const-rewriter

Reply:
This sounds similar to what we do with `apphost` today for the app dll and bundle marker (except it is the SDK at the user's build time that updates the binary):
https://github.com/dotnet/runtime/blob/cfe3d2d02d0a452caa3770c4f792ee438513603d/src/native/corehost/corehost.cpp#L39-L40
https://github.com/dotnet/runtime/blob/309185fd7d4d87f3e42e475ef57961159d0ce2d2/src/native/corehost/apphost/bundle_marker.cpp#L11-L14
https://github.com/dotnet/runtime/blob/cfe3d2d02d0a452caa3770c4f792ee438513603d/src/installer/managed/Microsoft.NET.HostModel/AppHost/HostWriter.cs#L53-L54
https://github.com/dotnet/runtime/blob/cfe3d2d02d0a452caa3770c4f792ee438513603d/src/installer/managed/Microsoft.NET.HostModel/AppHost/HostWriter.cs#L180-L185

Reply:
> it is not that big of a deal to have never used 1800 bytes in the binary

The original description was creating a 'compact delta representation', but if this compact representation isn't going to ultimately cause any space to be saved then we may as well stick with the original representation right?

I'm not trying to push us to get this super optimized, I'm actually wondering if I was premature in believing default offsets were necessary to meet our size goals. Originally I was thinking about encodings that were 8+ bytes per field, but if we assume there are < 1000 fields and we stored only a 2 byte offset per field that already fits in 2KB without tracking any default values or post-processing. However I don't know if we have any equivalent of the C++ offsetof() operator for C# data structures in NativeAOT.

If everyone likes this current approach I don't want to be the blocker, it certainly looks like it would work fine, I just wonder if there is a simpler option that doesn't need any special build steps or new tools.

Reply:
There is a cost difference between unusual bytes in the binary and bytes that are read/written during startup and sent as part of minidump payload. The cost of a byte that is read/written during startup and sent as part of minidump payload is much higher than the cost of a byte that just sits unused in the binary. In any case, the solution suggested by @lambdageek avoids the unused bytes in the binary, so we are all good here.

> if we assume there are < 1000 fields and we stored only a 2 byte offset per field that already fits in 2KB without

You also need names for the fields to make this self-describing. I do not think we want to maintain ever-growing list of name ordinals.

Reply:
> You also need names for the fields to make this self-describing. I do not think we want to maintain ever-growing list of name ordinals.

Did I misunderstand? The original description of the array was `static const int[] = { <unique sequence that identifies this array>, offsetof(TypeA, FieldX), offsetof(TypeA, FieldY), offsetof(TypeB, FieldZ), .... }`. How would we know that the first offset in the array represents FieldX and not FieldY if we aren't keeping track of the ordinals?

Reply:
I need to think a little bit about how this will work for cross-compilation. I think it still works - we can use fixed size data types for the magic payload and we can detect endianness based on the prefix.  That should provide enough info for the tool to form a model of the target platform's runtime and compute a delta or identify the contract that matches the output.

I think the build will have to produce some kind of correlated output (ie: a preprocessed header file or something) that has the _names_ of the entities in the object file blob.  But I think it's a solvable implementation problem.  (Worst case we also dump a massive nul-separated string into the object file.  But I would like to avoid that)

Reply:
> How would we know that the first offset in the array represents FieldX and not FieldY if we aren't keeping track of the ordinals?

So I think this will work in three phases:

1. In the first phase, the runtime build will compile a bunch of object files that contain the `magic, sizeof(typeX), offsetof(typeX, fieldA), sizeof(typeY), offsetof(typeY, fieldB)...` data blob.  (**New** It will also emit a text file containing a list of (contract, not source-level) type name and type/membername items in the same order as the data blob).  Note this object file's sole purpose is so that we get the C/C++/NativeAOT compiler to produce something targeting the target environment (ie: it's built by a real 32-bit LE arm linux toolchain, so it has offsets relevant to that platform).  We dont' actually link this object file into the final output binary of the runtime. We just scrape it to extract the useful info.
2. A tool (to be written) will consume the object file and the description at build time and produce a (logical) data type descriptor (ie: a fully expanded out type descriptor in a format yet to be specified).  If the layout is an exact match for a well-known descriptor (ie: we 'publish' a final 'net9.0 linux-arm32-le' descriptor into the runtime repo at some point near the end of the release) then we create a source file where the descriptor is just a reference to the well known name.   if the final descriptor doesn't match exactly a well known one, we emit a delta (ie: here's the baseline and here's the offsets that differ).   The output of this step is a C source file with the descriptor in whatever format physical format we decide to implement.
3. We compile the source file from step(2) and link it into the final built artifact of the runtime.

I will open a separate PR against this baseline spec to outline this part.